docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,106 @@
+title = "SIP 017 - Service Chaining"
+template = "main"
+date = "2024-02-21T23:00:00Z"
+
+---
+
+Summary: Allow HTTP requests to application components without going via the network
+
+Owner(s): ivan.towlson@fermyon.com
+
+Created: Feb 21, 2024
+
+# Requirements and scope
+
+In Spin 2.2, a component can make a HTTP request with only a path (e.g. `/api/accounts`).  Such a request will be routed to the same application, but this is done by Spin prepending the host where the application is running (e.g. `https://mysrsbizness.fermyon.app/api/accounts`).  The consequence of this is that requests go out to the network and come back in, including through load balancers etc.  This adds overhead, and in some environnments could count against quota or incur bandwidth charges.
+
+We have [prototyped](https://github.com/fermyon/spin/pull/2229) dispatching these requests directly to the component handling the route, but this was in the context of an implementation detail of Spin.  The requirement for this project is for an application to declare that it _requires_ dispatch to run locally on the node, and for hosts that cannot meet that prerequisite to reject the application.  This allows developers to write applications as a network of chained microservices, without having to be concerned about what happens if the microservices are not co-located.
+
+# User experience
+
+## Expressing a requirement for chaining
+
+To express that the application requires chaining for self-requests, add the following to `spin.toml`:
+
+```toml
+[application.prerequisites]
+service_chaining = "required"
+```
+
+**BIKESHED:** Or something more explicit such as `direct_self_requests`.  Or `prerequisites` could be an array of capabilities that the host must provide (`prerequisites = ["service-chaining"]`).  Also, we _could_ call this something like `environment` or `host` instead, but I'm kind of inclined to leave it generic as to what prerequisites might go in here.
+
+Entries in `prerequisites` are considered must-understand.  For the section itself, that has to happen at a `serde` level since we can't retrofit it to existing builds.  But `serde` errors are not very lovely, and I'd prefer hosts in future to be able to say "hey I don't support your `gpu` requirement" rather than "unknown field serialisation error."
+
+## Requesting a component via chaining
+
+There are two ways we could express a HTTP request to a chained component:
+
+1. Use the existing self-request style.  The host _may_ chain self-requests, and _must_ chain them if the "require chaining" setting is on.  This is super easy for the application developer, at both the send and receive end.  However, it becomes more complex for the host, as the host now has to route the request - this drags in much more of the HTTP trigger plumbing.  (The prototype did this and it was egregiously hairy.)
```

Comment:
It also requires the developer to keep these requests in sync with the trigger routing table, which may not always be desirable.